src/hal/drivers/mesa-hostmot2/hm2_rpspi.c

```diff
@@ -0,0 +1,568 @@
+	uint8_t spibpf;	//bits per frame
+};
+
+static hm2_rpspi_t boards[MAX_BOARDS];
+static int comp_id;
+
+
+/*************************************************/
+//so untested, you could implement rtapi_open_as_root in "C" as
+int rtapi_open_as_root(const char *filename, int mode) {
+		setfsuid(geteuid());
+	int r = open(filename, mode);
+	setfsuid(getuid());
+	return r;
+}
+
+/*************************************************/
+// aib (address increment bit)
+static uint32_t mk_read_cmd(uint16_t addr, unsigned msglen, bool aib) {
+    return 0 | (addr << 16) | 0xA000 | (aib ? 0x800 : 0) | (msglen << 4);
+}
+
+/*************************************************/
+static uint32_t mk_write_cmd(uint16_t addr, unsigned msglen, bool aib) {
+    return 0 | (addr << 16) | 0xB000 | (aib ? 0x800 : 0) | (msglen << 4);
+}
+
+/*************************************************/
+/*
+SPI_CS_RXF		0x00100000	
+	RXF - RX FIFO Full
+		0 = RXFIFO is not full.
+		1 = RX FIFO is full. No further serial data will be sent/ received until data is read from FIFO.
+
+SPI_CS_RXR		0x00080000
+	RXR RX FIFO needs Reading ( full)
+		0 = RX FIFO is less than full (or not active TA = 0).
+		1 = RX FIFO is or more full. Cleared by reading sufficient data from the RX FIFO or setting TA to	0.
+
+SPI_CS_TXD		0x00040000
+	TXD TX FIFO can accept Data
+		0 = TX FIFO is full and so cannot accept more data.
+		1 = TX FIFO has space for at least 1 byte.
+
+SPI_CS_RXD		0x00020000
+	RXD RX FIFO contains Data
+		0 = RX FIFO is empty.
+		1 = RX FIFO contains at least 1 byte.
+*/
+static int hm2_rpspi_write(hm2_lowlevel_io_t *llio, uint32_t addr, void *buffer, int size) {
+	//rtapi_print_msg(RTAPI_MSG_ERR, "hm2_rpspi_write addr: %08x, size: %d", addr, size);
+    hm2_rpspi_t *this = (hm2_rpspi_t*) llio;
+    if(size == 0) return 0;
+    if(size % 4) return -EINVAL;
+
+    int msgsize = size/4;
+	uint8_t *tbuff;
+    uint8_t *buffer8 = (uint8_t *)buffer;
+	uint32_t *buffer32 = (uint32_t *)buffer;
+    int i=0, j=0;
+	uint8_t  gully;
+
+    this->txBuf[0] = mk_write_cmd(addr, msgsize, true);
+
+	 for(i=0; i<msgsize; i++)
+		this->txBuf[i+1] = buffer32[i];
+	
+	/* activate transfer and clear fifos */
+	//BCM2835_SPICS = SPI_CS_TA|SPI_CS_CPHA|SPI_CS_CLEAR_TX|SPI_CS_CLEAR_RX;
+	BCM2835_SPICS |= SPI_CS_TA|SPI_CS_CLEAR_TX|SPI_CS_CLEAR_RX;
+
+	tbuff = (uint8_t *)this->txBuf;
+
+	/* send txBuf */
+	/* RPi SPI transfers 8 bits at a time only */
+	for (i=0; i<msgsize+1; i++) {
+		for(j=3; j>=0; j--){
+			BCM2835_SPIFIFO = tbuff[j];
+			if(!(BCM2835_SPICS & SPI_CS_TXD))
+				rtapi_print_msg(RTAPI_MSG_ERR, "TX-FIFO is full!\n");
+		}
+		/* wait until transfer is finished */
+		while (!(BCM2835_SPICS & SPI_CS_DONE));
+		
+		if((BCM2835_SPICS & SPI_CS_RXR) || (BCM2835_SPICS & SPI_CS_RXF))
+			rtapi_print_msg(RTAPI_MSG_ERR, "RX-FIFO is full!\n");
+		
+		for(j=3; j>=0; j--){
+			if(!(BCM2835_SPICS & SPI_CS_RXD))
+				rtapi_print_msg(RTAPI_MSG_ERR, "RX-FIFO is empty!\n");
+			gully = BCM2835_SPIFIFO;
```

Comment:
I nearly removed this line (and the declaration of gully, above)  because buildbot says:
~~~~
hal/drivers/mesa-hostmot2/hm2_rpspi.c:157:11: warning: variable ‘gully’ set but not used [-Wunused-but-set-variable]
hal/drivers/mesa-hostmot2/hm2_rpspi.c:187:4: warning: Value stored to 'gully' is never read
~~~~
but just in time I realized that it is accessing volatile storage, and the side effect is what is important.

If my understanding is correct, then, consider writing
~~~~
(void)BCM2835_SPIFIFO; // discard one byte from read fifo
~~~~
as a clue to future developers who may also see these compiler diagnostics and want to take action.